internal/service/k8s/ipam.go

```diff
@@ -0,0 +1,309 @@
+		Name:      fmt.Sprintf("%s-%s", nic, suffix),
+	}
+
+	claim, err := h.GetIPAddressClaim(ctx, key)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return "", err
+		}
+		log.V(4).Info("IPAddressClaim not found, creating it.", "nic", nic)
+		err = h.CreateIPAddressClaim(ctx, machineScope.IonosMachine, key.Name, machineScope.ClusterScope.Cluster.Name, poolRef)
+		if err != nil {
+			return "", errors.Join(err, fmt.Errorf("unable to create IPAddressClaim for machine %s", machineScope.IonosMachine.Name))
+		}
+		// we just created the claim, so we can return early and wait for the creation of the IPAddress.
+		return "", nil
+	}
+
+	// we found a claim, lets see if there is an IPAddress
+	ipAddrName := claim.Status.AddressRef.Name
+	if ipAddrName == "" {
+		log.V(4).Info("No IPAddress found yet.", "nic", nic)
+		return "", nil
+	}
+
+	ipAddrKey := types.NamespacedName{
+		Namespace: machineScope.IonosMachine.Namespace,
+		Name:      ipAddrName,
+	}
+	ipAddr, err := h.GetIPAddress(ctx, ipAddrKey)
+	if err != nil {
+		return "", errors.Join(err, fmt.Errorf("unable to get IPAddress specified in claim %s", claim.Name))
+	}
+
+	ip = ipAddr.Spec.Address
+
+	log.V(4).Info("IPAddress found, ", "ip", ip, "nic", nic)
+
+	return ip, nil
+}
+
+// CreateIPAddressClaim creates an IPAddressClaim for a given object.
+func (h *Helper) CreateIPAddressClaim(ctx context.Context, owner client.Object, name string, cluster string, poolRef *corev1.TypedLocalObjectReference) error {
+	claimRef := types.NamespacedName{
+		Namespace: owner.GetNamespace(),
+		Name:      name,
+	}
+	ipAddrClaim := &ipamv1.IPAddressClaim{}
+	var err error
+	if err = h.client.Get(ctx, claimRef, ipAddrClaim); err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if !apierrors.IsNotFound(err) {
+		// IPAddressClaim already exists
+		return nil
+	}
+
+	desired := &ipamv1.IPAddressClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      claimRef.Name,
+			Namespace: claimRef.Namespace,
+			Labels:    map[string]string{clusterv1.ClusterNameLabel: cluster},
+		},
+		Spec: ipamv1.IPAddressClaimSpec{
+			PoolRef: *poolRef,
+		},
+	}
+	_, err = controllerutil.CreateOrUpdate(ctx, h.client, desired, func() error {
+		controllerutil.AddFinalizer(desired, infrav1.MachineFinalizer)
+		return controllerutil.SetControllerReference(owner, desired, h.client.Scheme())
+	})
+
+	return err
+}
+
+// GetIPAddress attempts to retrieve the IPAddress.
+func (h *Helper) GetIPAddress(ctx context.Context, key client.ObjectKey) (*ipamv1.IPAddress, error) {
+	out := &ipamv1.IPAddress{}
+	err := h.client.Get(ctx, key, out)
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// GetIPAddressClaim attempts to retrieve the IPAddressClaim.
+func (h *Helper) GetIPAddressClaim(ctx context.Context, key client.ObjectKey) (*ipamv1.IPAddressClaim, error) {
+	out := &ipamv1.IPAddressClaim{}
+	err := h.client.Get(ctx, key, out)
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
```

Comment:
If you want you could also do it like this (untested):

```suggestion
func getResource[T client.Object](ctx context.Context, key client.ObjectKey) (*T, error) {
        var out T
	err := h.client.Get(ctx, key, &out)
	if err != nil {
		return nil, err
	}

	return &out, nil
}

func (h *Helper) GetIPAddress(ctx context.Context, key client.ObjectKey) (*ipamv1.IPAddress, error) {
      return getResource[ipamv1.IPAddress](ctx, key)
}

// GetIPAddressClaim attempts to retrieve the IPAddressClaim.
func (h *Helper) GetIPAddressClaim(ctx context.Context, key client.ObjectKey) (*ipamv1.IPAddressClaim, error) {
      return getResource[ipamv1.IPAddressClaim](ctx, key)
}
```

Reply:
This gives: Cannot use ipamv1.IPAddressClaim as the type client.Object Type does not implement 'client.Object' as the 'DeepCopyObject' method has a pointer receiver.

I tried a bit, but did not find a good solution.